posts/MarcoB/2024-07-12-embeddable-inheritance.adoc

```diff
@@ -0,0 +1,237 @@
+
+image::../../images/marco/embeddable-inheritance.png[]
+
+When this is the case, the `@Embedded` properties using that type will create an *additional column* in the entity mappings in which they are contained that will store information about the composite value's specific subtype. When retrieving the inherited property, Hibernate will read the discriminator value and instantiate the correct @Embeddable subtype with its corresponding properties.
+
+Here's an example of how you can enable embeddable inheritance in your mappings. Taking from the <<embeddable-values,previous chapter's example>>, let's say the `Animal` type is extended by other embeddables:
+====
+[source, java, indent=0]
+----
+@Embeddable
+class Mammal extends Animal {
+	private String mother;
+}
+
+@Embeddable
+class Cat extends Mammal {
+	// [...]
+}
+
+@Embeddable
+class Dog extends Mammal {
+	// [...]
+}
+
+@Embeddable
+class Fish extends Animal {
+	private int fins;
+}
+----
+====
+
+This will enable embeddable inheritance and, on the corresponding `Owner` entity, we'll see the columns for all the subtypes properties as well as the discriminator:
+====
+[source, sql, indent=0]
+----
+    create table Owner (
+        id bigint not null,
+        pet_DTYPE varchar(31) not null,
+        age integer,
+        name varchar(255),
+        mother varchar(255),
+        fins integer,
+        primary key (id)
+    )
+----
+====
+
+[NOTE]
+====
+Embeddable inheritance *IS* also supported for components used in an `@ElementCollection`.
+Embeddable inheritance *is NOT* supported for `@EmbeddedId`, embeddable types used as `@IdClass`
+and embedded properties using a custom `@CompositeType`.
+====
+
+==== Why Use Embeddable Inheritance?
+
+Embeddable inheritance is useful when you want to embed a polymorphic structure within your mapped tables, instead of relying on the entire entity mapping itself to handle inheritance. It allows for a clean and modular design, enabling code reuse and maintaining a clear separation of concerns.
+
+==== Customizing the discriminator column
+
+By default, the discriminator column will be STRING typed and named <property_name>_DTYPE, where property_name is the name of the @Embedded property attribute in the respective entity mapping. It's possible to customize the discriminator column mapping:
+
+* For the whole @Embeddable type, by using @DiscriminatorColumn or @DiscriminatorFormula on the root class of the inheritance hierarchy (NOTE: if using the same inheritance-enabled embeddable type for two different properties in the same entity mapping, this will cause a column name conflict);
+* For a specific @Embedded property, by using the @AttributeOverride annotation with a special value as name: `+{discriminator}+`.
```

Comment:
```suggestion
* For the whole `@Embeddable` type, by using `@DiscriminatorColumn` or `@DiscriminatorFormula` on the root class of the inheritance hierarchy (NOTE: if using the same inheritance-enabled embeddable type for two different properties in the same entity mapping, this will cause a column name conflict);
* For a specific `@Embedded` property, by using the `@AttributeOverride` annotation with a special value as name: `+{discriminator}+`.
```